### ISSUE TITLE
cli: Quit returns before DB shuts down

### ISSUE THREAD

**Issue by akshayjshah** [cockroachdb/cockroach#38851]
**Describe the problem**
I'm trying to reliably run multiple CockroachDB processes in automated tests (the same task led to #38758). This is harder than I expected, in part because `cockroach --quit` doesn't block until the server actually shuts down. This leads to a race: the server is shutting down, but my tests are already trying to start a new server using the same data directory. Often, the new process fails to start because the old process hasn't relinquished its locks:

```
* ERROR: could not cleanup temporary directories from record file: could not lock temporary directory /Users/ashah/cockroach-data/cockroach-temp659510233, may still be in use: IO error: While lock file: /Users/ashah/cockroach-data/cockroach-temp659510233/TEMP_DIR.LOCK: Resource temporarily unavailable
```

It's easy to work around this by sleeping and giving the first process a chance to shut down cleanly, but it takes a *very* long sleep (~10s) to make tests acceptably reliable.

It'd be much nicer if `quit` blocked until shutdown completes. We could change the default behavior or introduce a `--block` flag.

**To Reproduce**
This bash script fails reliably on my machine:
```bash
#!/usr/bin/env bash
set -euo pipefail

cockroach start --insecure --background
# Let the DB get its groove on.
sleep 1
cockroach quit --insecure
cockroach start --insecure --background
```

**Expected behavior**
I expected the `quit` subcommand to block until shutdown completes, either optionally or by default.

**Additional data / screenshots**
None.

**Environment:**
 - CockroachDB version: 19.1.1
```sh
$ cockroach version
Build Tag:    v19.1.1
Build Time:   2019/05/15 20:21:46
Distribution: CCL
Platform:     darwin amd64 (x86_64-apple-darwin14)
Go Version:   go1.11.6
C Compiler:   4.2.1 Compatible Clang 3.8.0 (tags/RELEASE_380/final)
Build SHA-1:  c8bda1de440cfe90cf23a433119d77795cfa0047
Build Type:   release
```
 - Server OS: MacOS
 - Client app: `cockroach` CLI

**Additional context**
This doesn't have any production impact, but working around this makes tests very slow.

I'm filing bugs about this because you're *so close* to a really amazing experience here. That doesn't take away from how much I appreciate what you've already done - integrating CockroachDB into my tests is already light-years better than MySQL. Keep up the great work! ❤️

**Comment by piyush-singh** [comment]
@akshayjshah thanks for filing! We'll take a look at this and think about how we can make the `quit` work a little better for your workflow.

cc @knz, we can chat about this on Wednesday

**Comment by knz** [comment]
@akshayjshah even if we were doing our best the solution would still not be to use `cockroach quit` as a race condition would always remain: the process closes its network connection (and thus `cockroach quit` stops being useful) before it closes its data directory, so there's always this little amount of time where you don't exactly know what is going on.

Since this is a test and you have access to the system where the server is running, a very reliable way to wait is to use `--pid-file` when starting up the server, and then use this loop (for example) in a shell script:

`while kill -CONT $(cat pidfile); do sleep .25; done`

(`kill -CONT` is a no-op as long as the process exists, and fails when the process has terminated)

Would this help?

**Comment by akshayjshah** [comment]
> @akshayjshah even if we were doing our best the solution would still not be to use cockroach quit as a race condition would always remain: the process closes its network connection (and thus cockroach quit stops being useful) before it closes its data directory, so there's always this little amount of time where you don't exactly know what is going on.

I don't think that's true, at least in many cases. In my case, my code doesn't have any races. I know that the tests are complete and no other code is trying to use the database when I begin shutting down the database; I just need to know when the database has completed shutdown so that I can start running the next batch of tests.

> Since this is a test and you have access to the system where the server is running, a very reliable way to wait is to use --pid-file when starting up the server, and then use this loop (for example) in a shell script...

Sure, that could work. It's pretty irritating, though - instead of using the CLI directly, between this issue and #38758, I'm in a wait loop checking files on both startup and shutdown.

**Comment by petermattis** [comment]
> the process closes its network connection (and thus cockroach quit stops being useful) before it closes its data directory, so there's always this little amount of time where you don't exactly know what is going on.

@knz We could change `cockroach quit` to have another step where we stop processing connections (while still accepting them), close the data directory (releasing the file lock), and then exiting the process. Do you see any gaps in that approach?

> Sure, that could work. It's pretty irritating, though - instead of using the CLI directly, between this issue and #38758, I'm in a wait loop checking files on both startup and shutdown.

@akshayjshah Do you have to reuse the same database for these tests? The common approach for testing I've used in the past is to have each test create a fresh database from scratch.

**Comment by akshayjshah** [comment]
> @akshayjshah Do you have to reuse the same database for these tests? The common approach for testing I've used in the past is to have each test create a fresh database from scratch.

Most of my unit tests do exactly what you describe. For running the whole application locally, though, it's convenient to keep the data in a known location (so changing a little code and restarting the local application doesn't blow away all the data). The unit tests for *that* code exercise starting and stopping the same database, and run right into this issue. It's not a really big deal, but it's another small surprise in a CLI that's otherwise pretty wonderful.

**Comment by knz** [comment]
> @knz We could change cockroach quit to have another step where we stop processing connections (while still accepting them), close the data directory (releasing the file lock), and then exiting the process. Do you see any gaps in that approach?

Yes architecturally this is possible. We'd need some plumbing to get there though. I'll see with Piyush about prioritizing the work.

**Comment by knz** [comment]
@petermattis note however that your comment was not well-formed: it is not the task of `cockroach quit` to close the process. I took your comment to mean that the `cockroach` server process must keep its client socket to `cockroach quit` open during the entire shutdown process, up to and including after all the storage/kv/rpc machinery is shut down (as all of it requires gossip/storage to be up). 

Keeping that socket around for so long will violate a couple of encapsulations but if the ends justify the means, so be it.

**Comment by tbg** [comment]
FYI we discussed elsewhere getting rid of ./cockroach quit or at least
relying on it less.

**Comment by knz** [comment]
To quote the powers that be:

> Ben's point of view is that whatever orchestrates your deployment is in charge of that. For example, if you use quit against a node running under k8s, k8s will respawn it. The moment some system is in charge of process lifecycle, that system needs to be in charge of.. process lifecycle. If you can start a process, you can also send it a signal.

@akshayjshah I think the lesson here is that perhaps your tests would be best served by having the process that starts `cockroach` also be responsible for stopping the server (e.g. by means of a signal) and simply wait on it to terminate.

**Comment by akshayjshah** [comment]
Fair enough. If you'd prefer not to change the behavior of `cockroach quit`, feel free to resolve this issue.

Thanks for the discussion!

**Comment by knz** [comment]
We've addressed

- a new command `cockroach node drain` will make the server shut down but let the process run
- then the operator can terminate and restart the process using their service manager

This way the service/orchestration manager is in full control of the process. The `start` command should only be listed as the service's initialization command.
